From: Catalin Marinas <catalin.marinas@arm.com>
Subject: mm-huge_memory-initialise-the-tags-of-the-huge-zero-folio-fix
Date: Mon, 3 Nov 2025 14:30:12 +0000

Handle the arch mte_zero_clear_page_tags() code issuing MTE
instructions irrespective of whether the hardware supports it.

Link: https://lkml.kernel.org/r/aQi8dA_QpXM8XqrE@arm.com
Signed-off-by: Catalin Marinas <catalin.marinas@arm.com>
Reported-by: Mark Brown <broonie@kernel.org>
Closes: https://lkml.kernel.org/r/aQiu-hU3x5V5drHF@finisterre.sirena.org.uk
Tested-by: Beleswar Padhi <b-padhi@ti.com>
Cc: Aishwarya TCV <aishwarya.tcv@arm.com>
Cc: Catalin Marinas <catalin.marinas@arm.com>
Cc: Mark Brown <broonie@kernel.org>
Cc: Will Deacon <will@kernel.org>
Cc: David Hildenbrand (Red Hat) <david@kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 arch/arm64/mm/fault.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/arch/arm64/mm/fault.c~mm-huge_memory-initialise-the-tags-of-the-huge-zero-folio-fix
+++ a/arch/arm64/mm/fault.c
@@ -969,6 +969,16 @@ struct folio *vma_alloc_zeroed_movable_f
 
 void tag_clear_highpage(struct page *page)
 {
+	/*
+	 * Check if MTE is supported and fall back to clear_highpage().
+	 * get_huge_zero_folio() unconditionally passes __GFP_ZEROTAGS and
+	 * post_alloc_hook() will invoke tag_clear_highpage().
+	 */
+	if (!system_supports_mte()) {
+		clear_highpage(page);
+		return;
+	}
+
 	/* Newly allocated page, shouldn't have been tagged yet */
 	WARN_ON_ONCE(!try_page_mte_tagging(page));
 	mte_zero_clear_page_tags(page_address(page));
_
